node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/IP.java

```diff
@@ -51,48 +52,140 @@
         return 0;
     };
 
-    /** A pool of available IP addresses */
-    public static class AddressPool {
+    /** IP configuration of a node */
+    public static class Config {
+
+        public static final Config EMPTY = new Config(Set.of(), Set.of());
+
+        private final Set<String> primary;
+        private final Pool pool;
+
+        /** DO NOT USE in non-test code. Public for serialization purposes. */
+        public Config(Set<String> primary, Set<String> pool) {
+            this.primary = ImmutableSet.copyOf(Objects.requireNonNull(primary, "primary must be non-null"));
+            this.pool = new Pool(Objects.requireNonNull(pool, "pool must be non-null"));
+        }
+
+        /** The primary addresses of this. These addresses are used when communicating with the node itself */
+        public Set<String> primary() {
+            return primary;
+        }
+
+        /** Returns the IP address pool available on a node */
+        public Pool pool() {
+            return pool;
+        }
+
+        /** Returns a copy of this with pool set to given value */
+        public Config with(Pool pool) {
+            return new Config(primary, pool.asSet());
+        }
+
+        /** Returns a copy of this with pool set to given value */
+        public Config with(Set<String> primary) {
+            return new Config(require(primary), pool.asSet());
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Config config = (Config) o;
+            return primary.equals(config.primary) &&
+                   pool.equals(config.pool);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(primary, pool);
+        }
+
+        @Override
+        public String toString() {
+            return String.format("ip config primary=%s pool=%s", primary, pool.asSet());
+        }
+
+        /** Validates and returns the given addresses */
+        public static Set<String> require(Set<String> addresses) {
+            try {
+                addresses.forEach(InetAddresses::forString);
+            } catch (IllegalArgumentException e) {
+                throw new IllegalArgumentException("Found one or more invalid addresses in " + addresses, e);
+            }
+            return addresses;
+        }
+
+        /**
+         * Verify IP config of given nodes
+         *
+         * @throws IllegalArgumentException if there are IP conflicts with existing nodes
+         */
+        public static List<Node> verify(List<Node> nodes, LockedNodeList allNodes) {
```

Comment:
This needs an additional check: If the `node` is a parent and the `other` is its child, then node's pool should be allowed to have `other`'s primaries. 

In fact, that in it self should maybe be a requirement, that all primaries of a child are in the pool of its parent, but there are some config servers and controllers that do not meet this criteria right now and due to this bug, its not possible to fix it.

Reply:
I was fixing this while working on an unrelated problem, but the fix wrecked all the tests. I'll just leave this here in case I revisit it later:
```
public static List<Node> verify(List<Node> nodes, LockedNodeList allNodes) {
    // Combine the new nodes with the already existing nodes, if the new node already exists
    // (when this is called as part of update of a node) use the updated version as we want
    // to verify that the state _after_ the update is valid
    Map<String, Node> nodesByHostname = nodes.stream()
            .collect(Collectors.toMap(Node::hostname, Function.identity()));
    NodeList updatedAllNodes = allNodes.asList().stream()
            .map(node -> nodesByHostname.getOrDefault(node.hostname(), node))
            .collect(Collectors.collectingAndThen(Collectors.toList(), NodeList::new));

    // We must ensure that:
    // 1. Child nodes do not have any addresses in the pool
    // 2. Set of host's primaries and set of host's addresses in the pool are disjoint
    // 3. All addresses assigned to a docker host (host primaries + pool) are disjoint from all other addresses
    //    assigned to all other hosts
    // 4. All child's primaries are a subset if host's pool
    for (Node node : nodes) {
        Optional<Node> parent = updatedAllNodes.parentOf(node);
        Set<String> addresses = new HashSet<>(node.ipConfig().primary());
        if (node.type().isDockerHost()) {
            addresses.addAll(node.ipConfig().pool().asSet());
        } else {
            // Verify 1:
            if (!node.ipConfig().pool().asSet().isEmpty())
                throw new IllegalArgumentException(node.hostname() + " (" + node.type() +
                        ") cannot have additional IP addresses");
        }

        if (parent.isEmpty()) { // TODO: Combine with the previous if when BM tenant nodes are gone
            // Verify 2:
            Set<String> hostOverlap = findOverlap(node.ipAddresses(), node.ipAddressPool().asSet());
            if (! hostOverlap.isEmpty())
                throw new IllegalArgumentException("Cannot assign " + addresses + " to " + node.hostname() +
                        ": " + hostOverlap + " are both in IP addresses and additional IP addresses");

            // Verify 3:
            for (Node other : updatedAllNodes.parents()) {
                Set<String> otherAddresses = new HashSet<>(other.ipConfig().primary());
                addresses.addAll(node.ipConfig().pool().asSet());

                Set<String> overlap = findOverlap(addresses, otherAddresses);
                if (! overlap.isEmpty())
                    throw new IllegalArgumentException("Cannot assign " + addresses + " to " + node.hostname() +
                            ": " + overlap + " already assigned to " + other.hostname());
            }
        } else {
            addresses.removeAll(parent.get().ipAddressPool().asSet());
            // Verify 4:
            if (!addresses.isEmpty())
                throw new IllegalArgumentException("Cannot assign " + node.ipConfig().primary() + " to " + 
                        node.hostname() + ": " + addresses + " are in the pool of " + parent.get().hostname());

        }


    }
    return nodes;
}

private static Set<String> findOverlap(Set<String> a, Set<String> b) {
    return a.stream()
            .filter(b::contains)
            .collect(Collectors.toUnmodifiableSet());
}
```